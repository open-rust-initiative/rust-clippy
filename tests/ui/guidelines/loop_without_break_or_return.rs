#![allow(unused, clippy::never_loop)]
#![warn(clippy::loop_without_break_or_return)]

fn test_01() {
    loop {
        println!("Hello, Rust!");
    } // lint

    loop {
        break;
    } // don't lint

    'outer: loop {
        break 'outer;
    } // don't lint

    'outer: loop {
        break;
    } // don't lint
}

fn test_02() {
    loop {
        if 2 < 3 {
            break;
        }
    } // don't lint
}

fn test_03() {
    'outer1: loop {
        for x in 0..5 {
            if x == 3 {
                break 'outer1;
            }
        }
    } // don't lint

    'outer2: loop {
        for x in 0..5 {
            if x == 3 {
                break;
            }
        } // lint
    }

    'outer3: loop {
        for x in 0..5 {
            if x == 3 {
                println!("Hello, Rust!");
            } else {
                break 'outer3;
            }
        }
    }
}

fn test_04() {
    'outer1: loop {
        loop {
            println!("Hello, Rust!");
        }
        break;
    }

    'outer2: loop {
        loop {
            break;
        }
    }

    'outer3: loop {
        loop {
            break 'outer3;
        }
    }

    'outer4: loop {
        'inner: loop {
            loop {
                break 'inner;
            }
        }
    }

    'outer5: loop {
        loop {
            'inner: loop {
                loop {
                    loop {
                        break 'inner;
                    }
                    break 'outer5;
                }
            }
        }
    }
}

fn test_05() {
    fn immediate_ret() {
        loop {
            return;
        }
    }

    fn ret_in_inner() {
        'outer: loop {
            // don't lint
            'inner: loop {
                return;
            }
        }
    }

    fn cond_ret_in_inner(flag: bool) {
        'outer: loop {
            // don't lint
            'inner: loop {
                if flag {
                    return;
                }
            }
        }
    }

    fn match_ret() {
        let a = Some(0_u8);
        loop {
            match a {
                Some(v) => (),
                None => return,
            }
        }
    }
}

macro_rules! set_or_ret {
    ($opt:expr, $a:expr) => {{
        match $opt {
            Some(val) => $a = val,
            None => return,
        }
    }};
}

fn ret_in_macro(opt: Option<u8>) {
    let opt: Option<u8> = Some(1);
    let mut a: u8 = 0;
    loop {
        set_or_ret!(opt, a);
    } // don't lint
}

fn match_pat() {
    let result: Result<u8, ()> = Ok(1);
    loop {
        let val = match result {
            Ok(1) => 1 + 1,
            Ok(v) => v / 2,
            Err(_) => return,
        };
    } // don't lint

    loop {
        let Ok(val) = result else { return };
    } // don't lint

    loop {
        let Ok(val) = result.map(|v| 10) else {
            break
        }; // don't lint
    }
}

fn exhaustive_loop() {
    for i in 0..5 {
        println!("{i}");
    } // don't lint

    let mut x = 0;
    while x < 5 {
        println!("x");
        x += 1;
    } // don't lint
}

fn infinite_inner() {
    loop {
        loop {
            println!("x");
        } // lint
        break;
    }
}

fn main() {}
