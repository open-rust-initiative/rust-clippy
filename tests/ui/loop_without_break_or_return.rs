#![allow(unused, clippy::never_loop)]
#![warn(clippy::loop_without_break_or_return)]

fn test_01() {
    loop {
        println!("Hello, Rust!");
    }

    loop {
        break;
    }

    'outer: loop {
        break 'outer;
    }

    'outer: loop {
        break;
    }
}

fn test_02() {
    loop {
        if 2 < 3 {
            break;
        }
    }
}

fn test_03() {
    'outer1: loop {
        for x in 0..5 {
            if x == 3 {
                break 'outer1;
            }
        }
    }

    'outer2: loop {
        for x in 0..5 {
            if x == 3 {
                break;
            }
        }
    }

    'outer3: loop {
        for x in 0..5 {
            if x == 3 {
                println!("Hello, Rust!");
            } else {
                break 'outer3;
            }
        }
    }
}

fn test_04() {
    'outer1: loop {
        loop {
            println!("Hello, Rust!");
        }
        break;
    }

    'outer2: loop {
        loop {
            break;
        }
    }

    'outer3: loop {
        loop {
            break 'outer3;
        }
    }

    'outer4: loop {
        'inner: loop {
            loop {
                break 'inner;
            }
        }
    }

    'outer5: loop {
        loop {
            'inner: loop {
                loop {
                    loop {
                        break 'inner;
                    }
                    break 'outer5;
                }
            }
        }
    }
}

fn main() {
    // test code goes here
}
