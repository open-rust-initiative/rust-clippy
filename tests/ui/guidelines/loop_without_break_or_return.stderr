error: loop without break condition
  --> $DIR/loop_without_break_or_return.rs:5:5
   |
LL | /     loop {
LL | |         println!("Hello, Rust!");
LL | |     }
   | |_____^
   |
   = help: consider adding `break` or `return` statement in the loop block
   = note: `-D clippy::loop-without-break-or-return` implied by `-D warnings`

error: loop without break condition
  --> $DIR/loop_without_break_or_return.rs:39:5
   |
LL | /     'outer2: loop {
LL | |         for x in 0..5 {
LL | |             if x == 3 {
LL | |                 break;
LL | |             }
LL | |         }
LL | |     }
   | |_____^
   |
   = help: consider adding `break` or `return` statement in the loop block

error: loop without break condition
  --> $DIR/loop_without_break_or_return.rs:60:9
   |
LL | /         loop {
LL | |             println!("Hello, Rust!");
LL | |         }
   | |_________^
   |
   = help: consider adding `break` or `return` statement in the loop block

error: loop without break condition
  --> $DIR/loop_without_break_or_return.rs:66:5
   |
LL | /     'outer2: loop {
LL | |         loop {
LL | |             break;
LL | |         }
LL | |     }
   | |_____^
   |
   = help: consider adding `break` or `return` statement in the loop block

error: loop without break condition
  --> $DIR/loop_without_break_or_return.rs:78:5
   |
LL | /     'outer4: loop {
LL | |         'inner: loop {
LL | |             loop {
LL | |                 break 'inner;
LL | |             }
LL | |         }
LL | |     }
   | |_____^
   |
   = help: consider adding `break` or `return` statement in the loop block

error: loop without break condition
  --> $DIR/loop_without_break_or_return.rs:87:9
   |
LL | /         loop {
LL | |             'inner: loop {
LL | |                 loop {
LL | |                     loop {
...  |
LL | |             }
LL | |         }
   | |_________^
   |
   = help: consider adding `break` or `return` statement in the loop block

error: aborting due to 6 previous errors

